plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,758 @@
+                        g_debug("failed data checksum comparison, expected: %u, got: %u",
+                                fu_sum16_bytes(fu_chunk_get_bytes(chk)),
+                                chksum);
+                } else
+                        return TRUE;
+                g_debug("retry attempt: %d, max: %d", retry, DDC_RW_MAX_RETRY_CNT);
+	}
+        g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "maximum tries exceeded");
+        return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_run_isp(FuMediatekScalerDevice *self, guint16 chksum, GError **error)
+{
+	g_autoptr(GByteArray) cod_commit = g_byte_array_new();
+	const guint8 buf[2] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_COMMIT_FW};
+        const guint8 data[2] = {
+                (guint8)(chksum & 0xFF),
+                (guint8)(chksum >> 8)
+        };
+
+	g_byte_array_append(cod_commit, buf, sizeof(buf));
+	g_byte_array_append(cod_commit, data, sizeof(data));
+	return mediatek_scaler_device_ddc_write(self, cod_commit, error);
+}
+
+static gboolean
+fu_mediatek_scaler_device_commit_firmware(FuMediatekScalerDevice *self, GBytes *fw, GError **error)
+{
+        if(!(fu_mediatek_scaler_device_run_isp(self, fu_sum16_bytes(fw), error))){
+                g_prefix_error(error, "failed to commit firmware: ");
+                return FALSE;
+        }
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_isp_reboot(FuMediatekScalerDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) cod_reboot = g_byte_array_new();
+	const guint8 buf[2] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_REBOOT};
+
+	g_byte_array_append(cod_reboot, buf, sizeof(buf));
+	if(!(mediatek_scaler_device_ddc_write(self, cod_reboot, error))){
+                g_prefix_error(error, "failed to set isp reboot: ");
+                return FALSE;
+        }
+        return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_get_isp_status(FuMediatekScalerDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) cmdbuf = g_byte_array_new();
+	g_autoptr(GByteArray) readbuf = g_byte_array_new();
+	const guint8 cmd[] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_GET_ISP_MODE};
+
+	g_byte_array_append(cmdbuf, cmd, sizeof(cmd));
+	if (!mediatek_scaler_device_ddc_read(self, cmdbuf, readbuf, error)) {
+		g_prefix_error(error, "failed to get isp status: ");
+		return FALSE;
+	}
+	if (readbuf->len < 2) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid data length on read: 0x%u",
+			    readbuf->len);
+		return FALSE;
+	}
+	if (readbuf->data[2] != 2) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "incorrect isp status, expected: 0x%02X, got 0x%u",
+			    (guint)2,
+			    readbuf->data[2]);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_verify(FuMediatekScalerDevice *self, GError **error)
+{
+	/* (TODO) delay 60+ seconds last until the display is back */
```

Comment:
how to repeately check the display status (until device back) during a period of time in a better way?

Reply:
fu_device_retry?